policy/composer.go

```diff
@@ -156,6 +195,62 @@ func (opt *ruleComposerImpl) rewriteVariableName(ctx *cel.OptimizerContext) ast.
 	})
 }
 
+func (opt *ruleComposerImpl) maybeUnnestRule(ctx *cel.OptimizerContext, ruleExpr ast.Expr) ast.Expr {
+	// Split the expr into local variables based on expression depth within ternaries
+	ruleAST := ctx.NewAST(ruleExpr)
+	ruleNav := ast.NavigateAST(ruleAST)
+	// Branches are ordered from leaf to root via the ast.MatchDescendants call.
+	heights := ast.Heights(ruleAST)
+	branchMap := map[int64]int{}
+	branches := []ast.NavigableExpr{}
+	ast.MatchDescendants(ruleNav, func(e ast.NavigableExpr) bool {
+		// // Compute and record the height of the expression
+		// calculator.computeHeight(e)
+
+		// If the expression is a comprehension, then all branches captured previously that relate to
+		// the comprehension body should be removed from the list of candidate branches for unnesting.
+		if e.Kind() == ast.ComprehensionKind {
+			// This only removes branches from the map, but not from the list of branches.
+			removeIneligibleConditions(e, branchMap)
+			return false
+		}
+		// Otherwise, if the expression is not a call, don't include it.
```

Comment:
Ternaries is our primary concern, so this can be a future enhancement -- but do we eventually want to split up nested lists/maps and field selections?

Reply:
It's possible, but the worst offenders are definitely the ternaries and logical AND/OR

Reply:
I've expanded the nesting check to all calls, though I believe I need to sort unnest candidates by height still.